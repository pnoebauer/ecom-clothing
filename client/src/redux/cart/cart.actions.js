import { CartActionTypes } from './cart.types';

export const toggleCartHidden = () => {
	return {
		type: CartActionTypes.TOGGLE_CART_HIDDEN
	};
};

export const addItemToCart = (payload) => {
	return {
		type: CartActionTypes.ADD_ITEM,
		payload
	};
};

export const removeItemFromCart = (payload) => {
	return {
		type: CartActionTypes.REMOVE_ITEM,
		payload
	};
};

export const clearItemFromCart = (payload) => {
	return {
		type: CartActionTypes.CLEAR_ITEM,
		payload
	};
};

export const clearCart = () => {
	return {
		type: CartActionTypes.CLEAR_CART
	};
};