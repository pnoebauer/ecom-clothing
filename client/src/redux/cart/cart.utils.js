export const addItemAndSetQuantity = (items, addItem) => {
	const itemExists = items.find(item => item.id === addItem.id); //check if the added item is already in the cart

	//if it is not in the cart yet add it with quantity 1
	if (!itemExists) {
		return [...items, {...addItem, quantity: 1}];
	}

	//if it exists, find it in the array and increase quantity by 1
	else {
		return items.map(item => {
			if (item.id === addItem.id) {
				return {
					...item,
					quantity: item.quantity + 1,
				};
			} else {
				return item;
			}
		});
	}
};

export const removeItemFromCart = (items, removeItem) => {
	if (removeItem.quantity <= 1) {
		return clearItemFromCart(items, removeItem);
	} else {
		return items.map(item => {
			if (item.id === removeItem.id) {
				return {
					...item,
					quantity: item.quantity - 1,
				};
			} else {
				return item;
			}
		});
	}
};

export const clearItemFromCart = (items, clearItem) => {
	return items.filter(item => item.id !== clearItem.id);
};
