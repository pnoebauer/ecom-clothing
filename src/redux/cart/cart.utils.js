export const addItemAndSetQuantity = (items, addItem) => {
	const itemExists = items.find(item => item.id === addItem.id); //check if the added item is already in the cart

	//if it is not in the cart yet add it with quantity 1
	if(!itemExists) {
		return [...items, { ...addItem, quantity: 1 }];
	}

	//if it exists, find it in the array and increase quantity by 1
	else {
		return items.map(item => {
			if(item.id === addItem.id) {
				return {
							...item,
							quantity: item.quantity + 1
						}
			}
			else {
				return item;
			}
		});
	}
};

export const clearItemFromCart = (items, removeItem) => {
	return items.filter(item => item.id !== removeItem.id);
}