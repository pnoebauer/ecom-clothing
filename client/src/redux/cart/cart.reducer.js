import CartActionTypes from './cart.types';

import {addItemAndSetQuantity, clearItemFromCart, removeItemFromCart} from './cart.utils';

const initialState = {
	hidden: true,
	items: [],
};

const cartReducer = (state = initialState, action) => {
	switch (action.type) {
		case CartActionTypes.TOGGLE_CART_HIDDEN:
			return {
				...state,
				hidden: !state.hidden,
			};
		case CartActionTypes.ADD_ITEM:
			return {
				...state,
				items: addItemAndSetQuantity(state.items, action.payload),
			};
		case CartActionTypes.REMOVE_ITEM:
			return {
				...state,
				items: removeItemFromCart(state.items, action.payload),
			};
		case CartActionTypes.CLEAR_ITEM:
			return {
				...state,
				items: clearItemFromCart(state.items, action.payload),
			};
		case CartActionTypes.CLEAR_CART:
			return {
				...state,
				items: [],
			};
		case CartActionTypes.SET_CART_FROM_FIREBASE:
			return {
				...state,
				items: action.payload,
			};
		default:
			return state;
	}
};

export default cartReducer;
