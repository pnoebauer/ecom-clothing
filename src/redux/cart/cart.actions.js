import { UserActionTypes } from './cart.types';

export const toggleCartHidden = () => {
	return {
		type: UserActionTypes.TOGGLE_CART_HIDDEN
	};
};

export const addItemToCart = (payload) => {
	return {
		type: UserActionTypes.ADD_ITEM,
		payload
	};
};

export const clearItemFromCart = (payload) => {
	return {
		type: UserActionTypes.CLEAR_ITEM,
		payload
	};
};