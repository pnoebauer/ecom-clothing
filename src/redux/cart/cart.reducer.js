import { UserActionTypes } from './cart.types';

import { addItemAndSetQuantity } from './cart.utils';

const initialState = {
	hidden: true,
	items: []
};

const cartReducer = (state=initialState, action) => {
	switch (action.type) {
		case UserActionTypes.TOGGLE_CART_HIDDEN:
			return {
				...state,
				hidden: !state.hidden
			};
		case UserActionTypes.ADD_ITEM:
			return {
				...state,
				items: addItemAndSetQuantity(state.items, action.payload)
			};
		default:
			return state;
	}
}

export default cartReducer;


