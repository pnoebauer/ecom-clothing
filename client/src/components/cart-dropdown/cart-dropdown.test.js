import React from 'react';
import {shallow} from 'enzyme';

import {CartDropdown} from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';
import {CartDropdownButton, EmptyMessageContainer} from './cart-dropdown.styles';

import {toggleCartHidden} from '../../redux/cart/cart.actions';

describe('CartDropdown component', () => {
	let wrapper;
	let mockHistory;
	let mockDispatch;
	const mockCartItems = [{id: 1}, {id: 2}, {id: 3}];

	beforeEach(() => {
		mockHistory = {
			push: jest.fn(),
		};

		mockDispatch = jest.fn();

		const mockProps = {
			cartItems: mockCartItems,
			history: mockHistory,
			dispatch: mockDispatch,
		};

		wrapper = shallow(<CartDropdown {...mockProps} />);
	});

	test('render CartDropdown component', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('call history.push when the button is clicked', () => {
		// const result = wrapper.find(CartDropdownButton);
		// console.log(result.debug());

		wrapper.find(CartDropdownButton).simulate('click');
		expect(mockHistory.push).toHaveBeenCalled();
		expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
	});

	test('renders a CartItem component for each element in the provided cartItems array ', () => {
		expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
	});

	test('render EmptyMessageContainer if cartItems is empty', () => {
		const mockProps = {
			cartItems: [],
			history: mockHistory,
			dispatch: mockDispatch,
		};

		const newWrapper = shallow(<CartDropdown {...mockProps} />);
		// console.log(newWrapper.debug());
		expect(newWrapper.exists(EmptyMessageContainer)).toBe(true);
	});
});
