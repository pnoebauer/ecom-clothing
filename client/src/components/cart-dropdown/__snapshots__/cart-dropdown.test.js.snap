// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component render CartDropdown component 1`] = `
<cart-dropdownstyles__CartDropdownContainer>
  <cart-dropdownstyles__CartItemsContainer>
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </cart-dropdownstyles__CartItemsContainer>
  <cart-dropdownstyles__CartDropdownButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </cart-dropdownstyles__CartDropdownButton>
</cart-dropdownstyles__CartDropdownContainer>
`;
